### PR TITLE
Improve spreadsheet import flow

### PR DESCRIPTION
**Pull request by alex-Arc**
import

### CODE REVIEW

File: apps/client/src/common/api/db.ts [hunk 1]
```diff
@@ -181,6 +181,17 @@ export async function importSpreadsheetPreview(file: File, options: ImportMap):
   return response.data;
 }
 
+export async function getSpreadsheetWorksheetNames(file: File): Promise<string[]> {
```
Comment (cpvalente) [inline apps/client/src/common/api/db.ts:184]: I am unsure I follow this endpoint, we already upload the excel, could we not use that instead?

**Review by cpvalente — COMMENTED** [review]
remember to run fix the linter issues.
I am not sure about the structure here, I will take some time to look into it once we can merge

File: packages/utils/src/date-utils/isTimeString.ts [hunk 2]
```diff
@@ -14,6 +14,6 @@ export const isTimeString = (text: string): boolean => {
   // ([0-5]?\d)          # SS (required)
   // $                   # End of string
 
-  const regex = /^(?:(?:([01]?\d|2[0-3])[:,.])?([0-5]?\d)[:,.])?([0-5]?\d)$/;
+  const regex = /^(?:(?:([01]?\d|2[0-3])[:,.])?([0-5]?\d)[:,.])?([0-5]?\d)?(\s)?([APap][Mm])?$/;
```
Comment (cpvalente) [inline packages/utils/src/date-utils/isTimeString.ts:17]: 🥶

File: e2e/tests/features/207-spreadsheet-import.spec.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  const fileChooserPromise = page.waitForEvent('filechooser');
+  await page.getByRole('button', { name: 'Import from spreadsheet', exact: true }).click();
+  const fileChooser = await fileChooserPromise;
+  await fileChooser.setFiles(fileToUpload);
+
+  const worksheetOption1 = await page.locator('[id="event\\ schedule"]').selectOption('Sheet2');
+  expect(worksheetOption1).toEqual(['Sheet2']);
+
+  const worksheetOption2 = await page.locator('[id="event\\ schedule"]').selectOption('test');
+  expect(worksheetOption2).toEqual(['test']);
```
Comment (cpvalente) [inline e2e/tests/features/207-spreadsheet-import.spec.ts:24]: really great effort here, I wonder if we could add a `data-testId` to make this selectors simpler?

File: apps/server/src/utils/time.ts [hunk 4]
```diff
@@ -41,16 +41,24 @@ const parse = (valueAsString: string): number => {
  * @param {boolean} fillLeft - autofill left = hours / right = seconds
  * @returns {number} - time string in millis
  */
-export const forgivingStringToMillis = (value: string, fillLeft = true): number => {
+export const forgivingStringToMillis = (value: string, fillLeft: boolean = true): number => {
   let millis = 0;
 
+  //force lower case before AM/PM check
+  value = value.toLowerCase();
```
Comment (cpvalente) [inline apps/server/src/utils/time.ts:48]: we shouldnt be mutating the parameters in a function. That may lead us to hard to find bugs were the data is changing in unpredictable ways, we can just make a new variable 
`const lowerValue = value.toLowerCase();`

File: apps/server/src/utils/parser.ts [hunk 5]
```diff
@@ -194,7 +194,7 @@ export const parseExcel = (excelData: unknown[][], options?: Partial<ImportMap>)
         event.title = makeString(column, '');
         // if this is a block, we have nothing else to import
         if (event.type === SupportedEvent.Block) {
-          continue;
+          /* empty */
```
Comment (cpvalente) [inline apps/server/src/utils/parser.ts:197]: i think the reason the continue was here was to stop the execution of that block.
That in turn would make sure we dont continue parsing an element after we found out it was a block

Why is this being removed? maybe I am missing something?

If there is a reason to remove the continuewe should remove the conditional entirely

File: apps/server/src/api-data/sheets/sheets.controller.ts [hunk 6]
```diff
@@ -56,6 +57,16 @@ export async function revokeAuthentication(_req: Request, res: Response) {
   }
 }
 
+export async function getWorksheetNamesFromSheet(req: Request, res: Response) {
+  try {
+    const { sheetId } = req.params;
+    const { worksheetOptions } = await getWorksheetOptions(sheetId);
```
Comment (cpvalente) [inline apps/server/src/api-data/sheets/sheets.controller.ts:63]: if this function has a single return, I dont see why we need the overhead of wrapping it in an object? am i missing something?

File: apps/server/src/api-data/excel/excel.controller.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+import { Request, Response } from 'express';
+import { extname, join } from 'path';
+import { existsSync, renameSync } from 'fs';
+import { generateId } from 'ontime-utils';
+import { uploadsFolderPath } from '../../setup/index.js';
+import { getExcelWorksheets, handleMaybeExcel } from '../../utils/parser.js';
+
+export async function postExcel(req: Request, res: Response) {
```
Comment (cpvalente) [inline apps/server/src/api-data/excel/excel.controller.ts:13]: I dont really like the idea of having controllers accessing file system. although I know that we have them, we shouldnt make this worse

lets create a excel.service.ts file and push some of the code to do with filesystem there.
Also, I think we have plenty of usages for existSync and renameSync that we could start cleaning up, i reckon we may have reusable functions that do this, or should do if not

**Review by cpvalente — COMMENTED** [review]
there is a lot of code here
I made a few suggestions for cleanup, lets start with those and I will look into the details more carefully after

File: packages/utils/src/date-utils/isTimeString.ts (hunk 2 above)
Comment (alex-Arc) [reply packages/utils/src/date-utils/isTimeString.ts:17]: update the comment

**Review by cpvalente — APPROVED** [review]
🚀

**Comment by alex-Arc** [comment]
if you have a chance to skim it that would be great
especially the new service file https://github.com/cpvalente/ontime/blob/improve-spreadsheet-import-flow/apps/server/src/api-data/excel/excel.service.ts

**Comment by alex-Arc** [comment]
image

this doesn't fail localy

File: apps/server/src/utils/parser.ts [hunk 8]
```diff
@@ -250,11 +243,16 @@ export const parseExcel = (excelData: unknown[][], options?: Partial<ImportMap>)
     // if any data was found in row, push to array
     const keysFound = Object.keys(event).length + Object.keys(eventCustomFields).length;
     if (keysFound > 0) {
-      if (timerTypeIndex === null) {
-        event.timerType = TimerType.CountDown;
-        event.type = SupportedEvent.Event;
+      // if it is a Block type drop all other filed
+      if (event.type === SupportedEvent.Block) {
```
Comment (alex-Arc) [inline apps/server/src/utils/parser.ts:247]: cleans up the block event

File: apps/server/src/utils/parser.ts [hunk 9]
```diff
@@ -192,10 +189,6 @@ export const parseExcel = (excelData: unknown[][], options?: Partial<ImportMap>)
         }
       } else if (j === titleIndex) {
         event.title = makeString(column, '');
-        // if this is a block, we have nothing else to import
-        if (event.type === SupportedEvent.Block) {
```
[the remark this answers is not in the document]
Comment (alex-Arc) [reply apps/server/src/utils/parser.ts:196]: yeah

it tried testing out what this skips, and calling continue in an exhaustive else if stack doesn't change anything
and as the type might not be the first thing found all the timer event fields will be added to a block event

so i added the clean up after https://github.com/cpvalente/ontime/pull/839/files/9b459c1fe43fe7359291a83f0f15907d06330d82#r1536170551

**Pull request merged**